README.md
GITHUB messes with the text and newlines, please view in "raw" mode...

##########################################
### 1. TRX2TXT tool suite description: ###
##########################################
A suite of shell scripts to display a Bitcoin transaction in plain text, similiar to the Bitcoin core client or "www.blockchain.info" JSON output. 

Scripts are written to run on OpenBSD and OSX and Linux systems at the command shell (ksh, bash). Existing tools on the web (Bitcoin CLI tools and others), are written to work only with BASHv4. In this suite, all scripts are coded with the intention, to be (nearly) POSIX compliant. Tested on OpenBSD korn shell, MAC OSX BASHv3 and SuSE Linux BASHv4. 

Main program (shell script) to display contents of a transaction is "trx2txt.sh". It refers to:
   https://en.bitcoin.it/wiki/Protocol_specification#tx 
The TRX_IN and TRX_OUT parts are shown, with the decoded sig script and pk pubkey script details, and it's corresponding bitcoin address(es). 
Script "trx2txt.sh" has several command line options (just open with "-h" or "--help").

Code is not (yet?) written for best performance: there are many calls to shell functions (which always fork during shell execution). And also several external (Unix standard) programs are called. 

Readability: current version of programs are heavily commented, to be able to follow/understand the program’s logic. 


#######################
### file trx2txt.sh ###
#######################
The main script. You'll want to start here :-) 

Usecases:

./trx2txt.sh 
  Without parameters, the details of a sample transaction are shown, line by line.

./trx2txt.sh -r 
  pass the hex code of a RAW TRANSACTION as option, for example the output of:
  https://blockchain.info/de/rawtx/cc8a279b0736e6a2cc20b324acc5aa688b3af7b63bbb002f46f6573c1ad84408?format=hex
  which would give some lines of hexcode. Simply cut&paste, and add after the -r parameter:
  ./trx2txt.sh -r 010000000253603b3fdb9d5e10de2172305ff68f4b5227310ba6bd81d4e1bf60c0de6183...

./trx2txt.sh -t
  pass the hex code of a TRANSACTION HASH as option, e.g.: 
  ./trx2txt.sh -t cc8a279b0736e6a2cc20b324acc5aa688b3af7b63bbb002f46f6573c1ad84408
  The script will check for network connectivity, and will then try to download the raw trx data from the given trx number.

./trx2txt.sh -u 
  pass the hex code of an UNSIGNED RAW TRX as option.
  This is an advanced option, so if you created your unsigned raw trx (eventually with 'trx_create_sign.sh' of this suite), it can be double checked.

The program checks the parameters, checks the version of the shell (for cases where POSIX compliance can not be achieved, in particular with arrays), checks availability for all necessary sub-programs (openssl, awk, bc, ...) to decode a transaction, and then begins to break down the transactions.

./trx2txt.sh -v
with parameter -v, a more detailed view is provided, along the guidelines of the bitcoin wiki.

./trx2txt.sh -vv
with parameter -vv, a very detailed view is provided, including the decoding of the sig script and the public key script. Two png files are provided, which show the state machine. The script verification/decoding is supported by three helper scripts (description below): 
trx_in_sig_script.sh - to decode the TRX_IN script
trx_out_pk_script.sh - to decode the TRX_OUT script
base58check_enc.sh   - to convert the scripts addresses to "human readable" bitcoin addresses

to understand the particular steps to decode a transaction, run this command:
  cat trx2txt.sh | grep STEP -A 2 -B 1 
The displayed steps are derived from: https://en.bitcoin.it/wiki/Protocol_specification#tx


###########################
### 2. Supporting files ###
###########################
The supporting scripts are used by the main program (trx2txt.sh), but can also be used independantly, when provided with the correct parameters. Use the '-h' parameters to these scripts to explain. 


#################################
### file trx_in_sig_script.sh ###
#################################
This shell script tries to decode the TRX-IN sig script via it's internal simple state diagram. The way the OPCodes are processed follows the picture "trx_in_sig_state_machine.png". At anytime the script can be used "stand alone". Without parameters, it displays as per below, alternativly just provide a sig script as parameter (or as usual, use '-h'). 

   48: OP_DATA_0x48
   30: OP_LENGTH_0x30
   45: OP_LENGTH_0x45
   02: OP_INT_0x02
   21: OP_LENGTH_0x21 *** this is SIG R
       00A428348FF55B2B:59BC55DDACB1A00F
       4ECDABE282707BA5:185D39FE9CDF05D7
       F0
   02: OP_INT_0x02
   20: OP_LENGTH_0x20 *** this is SIG S
       74232DAE76965B63:11CEA2D9E5708A0F
       137F4EA2B0E36D08:18450C67C9BA259D
       
   01: OP_SIGHASHALL *** This terminates the ECDSA signature (ASN1-DER structure)
 
   21: OP_DATA_0x21
   02: OP_INT_0x02
       025F95E8A33556E9:D7311FA748E9434B
       333A4ECFB590C773:480A196DEAB0DEDE
       E1
* This terminates the Public Key (X9.63 COMPRESSED form)
* corresponding bitcoin address is:
1HTNtayFkoBV28wqDygMbW33qSnRjxAuNR 

#################################
### file trx_out_pk_script.sh ###
#################################
This shell script tries to decode the TRX-OUT public key script via it's internal simple state diagram. The way the OPCodes are processed follows the picture "trx_out_pk_state_machine.png". At anytime the script can be used "stand alone". Without parameters, it displays as per below, alternativly just provide a sig script as parameter (or as usual, use '-h'). 

76A9146AF1D17462C6146A8A61217E8648903ACD3335F188AC
   76: OP_DUP
   A9: OP_HASH160
   14: OP_Data14 (= decimal 20)
       6AF1D17462C6146A:8A61217E8648903A
       CD3335F1
   88: OP_EQUALVERIFY
   AC: OP_CHECKSIG
* This is a P2PKH script
6AF1D17462C6146A8A61217E8648903ACD3335F1

The string of the 20 bytes in the middle of this output is the hexadecimal representation of the Bitcoin address. To show the "human readable" chars, a third script is used:


###################################
### file trx_base58check_enc.sh ###
###################################
This shell script tries to decode the addresses in BITCOIN SCRIPTs. It reads the hex codes, does a base58 conversion, and displays the bitcoin address. At anytime the file can be used "stand alone". Without parameters, it displays as per below, alternativly just provide a sig script as parameter (or as usual, use '-h').

./trx_base58check_enc.sh
using 010966776006953D5567439E5E39F86A0D273BEE
 
4: add 0x00 or 0x05 [P2SH] at the beginning
5. sha256
445c7a8007a93d8733188288bb320a8fe2debd2ae1b47f0f50bc10bae845c094
6. another sha256
d61967f63c7dd183914a4ae452c9f6ad5d462ce3d277798075b107615c1a8a30
7. take first four Bytes from step 6 as checksum
d61967f63c7dd183914a4ae452c9f6ad5d462ce3d277798075b107615c1a8a30
8. append checksum from step 7 to the result from step4
00010966776006953D5567439E5E39F86A0D273BEEd61967f6
9. encode Base58
16UwLL9Risc3QfPqBUvKofHmBQ7wMtjvM 


#################################
### file testcases_trx2txt.sh ###
#################################
This shell script supports the development process, and verifies the script output(s).
1.) build sha256 checksums of the involved source code scripts
2.) runs script "trx2txt.sh" with different parameters and transactions
3.) sends output into a file, and calculate it’s SHA256 hash value
4.) compare checksums 

Most easily it is used like this:

  ./testcases_trx2txt.sh 

which runs all tests (time consuming). This can be easily compared on all platforms. When hash is equal on all platforms, code is ready to be uploaded to GITHUB (or similiar). 

  ./trx_testcases.sh -l
  "-l", a log file ("trx_testcases.sh.log") is created. All checks are performed.

  ./trx_testcases.sh -h
  "-h" displays a help text

  ./trx_testcases.sh 1|2|3|4|5|6|7|8|9
  [1-9] runs only the mentioned testcases, to get quicker results


###############################
### file trx_create_sign.sh ###
###############################
Advanced usage! If you understand the idea of cold storage, and creating unsigned raw transactions, this is for you! The script tries to create or sign a transaction. It is a basis for an online/cold storage combination. On the Internet connected system you would create a raw transaction, copy it to your USB stick, and run the same script file(s) on the cold standby machine to sign the trx. Then copy the signed trx back to USB, bring it back to the Internet connected machine, and send the trx to the network. 
At anytime the file can be used "stand alone". Obviously not very useful without parameters, just provide a sig script as parameter (or as usual, use '-h').

Usage example:
(You will need to know the first three parameters from the previous trx, from which you want to redeem. Start with '-h' to better undertsand. Also do not try this unless previous transaction is confirmed!)

  ./trx_create_sign.sh -c -m <trx hash> <output> <pubkey script> <amount> <address> 

translates into something like this:
  ./trx_create_sign.sh -c -m c3434be....5c5b7a310cc67 0 76A9141FE307887696CF781DA237DBE2E12DB05C10986A88AC 110000 12GTF5ARSrqJ2kZd4C9XyGPesoVgP5qCdM

Hint: in its initial version the create/sign process allows for one input, one output and uses a P2PKH structure.


#########################################
### file testcases_trx_create_sign.sh ###
#########################################
This shell script supports the development process, and verifies the script output(s).
It is setup the exactly same way as 'testcases_trx2txt.sh'. Details up there ...
Most easily it is used like this:

  ./testcases_trx_create_sign.sh 


###########################
### file trx_key2pem.sh ### 
###########################
This shell script is necessary when the signing process is executed. The signature is done using the 'openssl' suite, which requires PEM keys. This script helps to convert from wif, wif-c or hex to PEM. At anytime the file can be used "stand alone". Obviously not very useful without parameters, just provide a sig script as parameter (or as usual, use '-h').

Hint: in its initial version the create/sign process allows for one input, one output and uses a P2PKH structure.


#####################################
### file testcases_trx_key2pem.sh ###
#####################################
This shell script supports the development process, and verifies the script output(s).
It is setup the exactly same way as 'testcases_trx2txt.sh'. Details up there ...
Most easily it is used like this:

  ./testcases_trx_key2pem.sh 


######################################
### file trx_verify_bc_address.awk ###
######################################
This awk script is a little helper tool, to make the code be more transportable. It is used by 
trx_create_sign.sh and trx_key2pem.sh.


##################################
### file trx_verify_hexkey.awk ###
##################################
This awk script is a little helper tool, to make the code be more transportable. It is used by 
trx_create_sign.sh and trx_key2pem.sh.


###########################
### file trx_base58.awk ###
###########################
This awk script is a little helper tool, to make the code be more transportable. It is used by 
trx_create_sign.sh and trx_key2pem.sh.


#########################
### 3. Documentation: ###
#########################
README.md                        - this file :-)
Changelog.txt                    - view changes over the files 
trx_in_sig_state_machine.graphml - graphics source file for the state machine
                                   (java based app: "yEd Graph Editor")
trx_in_sig_state_machine.png     - the exported png for script sig part
trx_out_pk_state_machine.graphml - graphics source file for the state machine
                                   (java based app: "yEd Graph Editor")
trx_out_pk_state_machine.png     - the exported png for PUBKEY script
trx_OP_CODES_desc.txt            - used references for PK_SCRIPT and their OpCodes
trx_state_matrix.ods             - table with three sheets, showing the state machine’s logic
trx_todos.txt*

Documentation files are not included in the hashing with any testcases.




Changelog.txt
# CHANGELOG 
# all updates changes to the trx2txt.sh suite.
#
# help and references:
# https://blockchain.info/de/rawtx/<trx number>?format=hex
# https://en.bitcoin.it/wiki/Protocol_specification#tx
#
# 
date    by    file                   comment
yymmmdd
16jun10 svn   all :-)                initial release
16jun11 svn   trx_out_pk_script.sh   convert parameter "to upper" string before decoding
16jun11 svn   trx_out_pk_script.sh   fix state machine for OP_DATA33
16jun11 svn   trx2txt.sh             when TX_OUT is zero, don't call subsequent scripts, 
                                     that would display "wrong" default output
16jun11 svn   trx2txt.sh             do a base58check_enc only, if length = 20 bytes
16jun12 svn   trx_in_sig_script.sh   converted AWK tool to shell script, added address decoding
16jun13 svn   trx_in_sig_script.sh   fixed S6_Length procedure, to verify next OPCode
16jun13 svn   trx_in_sig_script.sh   added sanity checks, and adopted naming to bitcoin core,
                                     as in interpreter.cpp or h files
16jun13 svn   trx_in_sig_state_machine.graphml - updated status names accordingly
16jun13 svn   trx_in_sig_script.sh   added sanity checks, and adopted naming to bitcoin core,
                                     as in interpreter.cpp or h files
16jun14 svn   trx_in_sig_script.sh   check if "sed" is available
                                     changed SIG R & SIG S to be 0x21 or 0x21
16jun14 svn   trx_testcases.sh       added log file capability
16jun15 svn   trx2txt.sh             parameter handling changed completly ("-r", "-t")
16jun15 svn   trx2txt.sh             add feature to load data from www.blockchain.info
16jun15 svn   trx_testcases.sh       added checksum results into output
16jun16 svn   trx_testcases.sh       added better parameter handling ([1-8])
16jun16 svn   trx2txt.sh             fixed naming conventions ("TRANSACTION_ID")
16jun16 svn   trx_in_sig_script.sh   added procedures S21-S23 
16jun16 svn   trx_in_sig_state_machine.graphml - added S21 beginning with "OP_Data_0x49"
16jun17 svn   trx2txt.sh             if TRX_IN sig script len=0, don't call trx_in_sig_script.sh
16jun17 svn   trx_in_sig_state_machine.graphml - added S24 beginning with "OP_Data_0x3C"
16jun18 svn   trx_out_pk_script.sh   fix state machine for OP_DATA33 and OP_DATA65 (P2PK)
16jun18 svn   trx_out_pk_state_machine.graphml - updated S&_OP_Data65 and S11_OP_Data33 
16jun18 svn   base58check_enc.sh     rework parameter handling 
16jun18 svn   trx2txt.sh             adopted to fit parameter handling for base58check_enc.sh
16jun18 svn   trx_in_sig_script.sh   adopted to fit parameter handling for base58check_enc.sh
16jun19 svn   trx2txt.sh             fixing "var_int" or "compact size integers

16jun20 svn   trx2txt.sh             work on integrating unsigned raw transactions 
16jun21 svn   trx2txt.sh             included an awk call (reverse_hex), to reverse hex data 
16jun21 svn   trx_in_sig_state_machine.graphml - added S30-S34 for unsigned raw trx data   
16jun25 svn   trx_in_sig_script.sh   added proper anming to Sig and pubkey (ECDSA, ASN1-DER and X9.63)
16jun27 svn   trx2txt.sh             work on integrating signing a raw transaction 
16jul02 svn   trx2txt.sh             display properly unsigned raw trx
16jul07 svn   trx2txt.sh             extended parameter handling
16jul08 svn   trx_create_sign.sh     created new file to properly manage creating/signing trx
16jul09 svn   trx_key2pem.sh         created file to manage key handling (hex to PEM) for signing
16jul09 svn   trx-suite	             ### released version to GITHUB ###
16jul17 svn   trx2txt.sh             fixed var_int (V_INT) handling
16jul17 svn   trx_create_sign.sh     fixed parameter handling for '-u'
16jul27 svn   trx_in_sig_state_machine.graphml - added S35-S37 for MULTISIG
16jul27 svn   trx_in_sig_script.sh   added multisig handling 
16jul27 svn   trx2txt.sh             corrected code for unsigned raw trx
16jul28 svn   trx_create_sign.sh     included strict DER-encoded signature checks 
16jul30 svn   trx_in_sig_script.sh   rewrote code: use "cut -b" for "get_next_char"
                                     get rid of "if bash == BASH4", no array handling
16jul31 svn   trx2txt.sh             added error check when fetching from blockchain.info
16jul31 svn   trx_in_sig_script.sh   fixed minor issue, were wrong address was displayed
16jul31 svn   trx_in_sig_state_machine.graphml - removed unnecessary status, and added 
                                                 redeem script colors for Multisig at the top
16aug01 svn   trx_in_sig_script.sh   fixed multisig redeem script in "TX IN", to display
                                     correct type "3" addresses 





trx_out_pk_script.sh
#!/bin/sh
##############################################################################
# Read the bitcoin PK_script OPCODES from a transaction's TRX_OUT
# script by Sven-Volker Nowarra 
#
# Version	by	date	comment
# 0.1		svn	02jun16	initial release
# 
# Copyright (c) 2015, 2016 Volker Nowarra 
# Complete rewrite of code in June 2016 from following reference:
#   https://en.bitcoin.it/wiki/Protocol_specification#tx
#   https://en.bitcoin.it/wiki/Script
# 
#  As of Bitcoin Core 0.10, these are considered standard scripts:
# 
#   P2PKH (pay-to-public-key-hash)
#   P2SH (pay-to-script-hash)
#   P2PK (pay-to-public-key)
#   Multisignature
#   OP_RETURN metadata
# 
# 
# Permission to use, copy, modify, and distribute this software for any 
# purpose with or without fee is hereby granted, provided that the above 
# copyright notice and this permission notice appear in all copies. 
# 
# 
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES 
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF 
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY 
# SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER 
# RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, 
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE 
# USE OR PERFORMANCE OF THIS SOFTWARE. 
#

ret_string=''
QUIET=0
param=76A9146AF1D17462C6146A8A61217E8648903ACD3335F188AC

case "$1" in
  -q)
     QUIET=1
     shift
     ;;
  -?|-h|--help)
     echo "usage: trx_out_sig_script.sh [-?|-h|--help|-q] hex_string"
     echo "  "
     echo "convert a raw hex string from a bitcoin trx-out into it's OpCodes. "
     echo "if no parameter is given, the data from a demo trx is used. "
     echo "  "
     exit 0
     ;;
  *)
     ;;
esac

if [ $QUIET -eq 0 ] ; then 
  echo "###############################################################"
  echo "### trx_out_pk_script.sh: read PK_script OPCODES from a trx ###"
  echo "###############################################################"
  echo "  "
fi

if [ $# -eq 0 ] ; then 
  if [ $QUIET -eq 0 ] ; then 
    echo "no parameter, hence showing example pk_script:"
    echo "$param"
  fi
else 
  param=$( echo $1 | tr "[:lower:]" "[:upper:]" )
fi

###########################################################################
### procedure to show data following an "OP_DATA" opcode                ###
###########################################################################
op_data_show() {
  n=1
  output=
  while [ $n -le $cur_opcode_dec ]
   do
    output=$output${opcode_ar[offset]}
    ret_string=$ret_string${opcode_ar[offset]}
    if [ $n -eq 8 ] || [ $n -eq 24 ] || [ $n -eq 40 ] || [ $n -eq 56 ] ; then 
      output=$output":"
    fi
    if [ $n -eq 16 ] || [ $n -eq 32 ] || [ $n -eq 48 ] || [ $n -eq 64 ] ; then 
      echo "       $output" 
      output=
    fi
    n=$(( n + 1 ))
    offset=$(( offset + 1 ))
  done 
  echo "       $output" 
}

#####################
### GET NEXT CODE ###
#####################
get_next_opcode() {
  cur_opcode=${opcode_ar[offset]}
  cur_hexcode="0x"$cur_opcode
  cur_opcode_dec=$( echo "ibase=16;$cur_opcode" | bc )
  # echo "offset=$offset, opcode=$cur_opcode, opcode_dec=$cur_opcode_dec"
  offset=$(( offset + 1 ))
}

#####################################
### STATUS 1 (OP_DUP)             ###
#####################################
S1_OP_DUP() {
  get_next_opcode
  case $cur_opcode in
    A9) echo "   $cur_opcode: OP_HASH160"
        S2_OP_HASH160
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 2 (OP_HASH160)         ###
#####################################
S2_OP_HASH160() {
  get_next_opcode
  case $cur_opcode in
    14) echo "   $cur_opcode: OP_Data$cur_opcode (= decimal $cur_opcode_dec)"
        op_data_show
        S3_OP_DATA20
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 3 (OP_DATA20)          ###
#####################################
S3_OP_DATA20() {
  get_next_opcode
  case "$cur_opcode" in
    88) echo "   $cur_opcode: OP_EQUALVERIFY"
        S4_OP_EQUALVERIFY
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 4 (OP_EQUALVERIFY)     ###
#####################################
S4_OP_EQUALVERIFY() {
  get_next_opcode
  case $cur_opcode in
    AC) echo "   $cur_opcode: OP_CHECKSIG"
        S5_P2PKH
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 5 (P2PKH)              ###
#####################################
S5_P2PKH() {
  echo "* This is a P2PKH script"
}
#####################################
### STATUS 6 (OP_DATA65)          ###
#####################################
S6_OP_DATA65() {
  get_next_opcode
  case $cur_opcode in
    AC) echo "   $cur_opcode: OP_CHECKSIG"
        S7_P2PK
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 7 (P2PK)               ###
#####################################
S7_P2PK() {
    echo "* This is a P2PK script"
}
#####################################
### STATUS 8 (OP_HASH160)         ###
#####################################
S8_OP_HASH160() {
  get_next_opcode
  case $cur_opcode in
    14) echo "   $cur_opcode: OP_Data$cur_opcode (= decimal $cur_opcode_dec)"
        op_data_show
        S9_OP_DATA20
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 9 (OP_DATA20)          ###
#####################################
S9_OP_DATA20() {
  get_next_opcode
  case $cur_opcode in
    87) echo "   $cur_opcode: OP_EQUAL" 
        S10_P2SH
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 10 (P2SH)              ###
#####################################
S10_P2SH() {
    # P2SH addresses start with a "05" in their base58encoded representation!
    # https://en.bitcoin.it/wiki/List_of_address_prefixes
    echo "* This is a P2SH script"
    # we add a "P2SH" string at the end, to notify the following 
    # base58encode script about this special adress
    ret_string="$ret_string -P2SH"
}
#####################################
### STATUS 11 (OP_DATA33)         ###
#####################################
S11_OP_DATA33() {
  get_next_opcode
  case $cur_opcode in
    AC) echo "   $cur_opcode: OP_CHECKSIG"
        S12_P2PK
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 12 (P2PK)              ###
#####################################
S12_P2PK() {
    echo "* This is a P2PK script"
}
#####################################
### STATUS 13 (OP_1)              ###
#####################################
S13_OP_1() {
  get_next_opcode
  case $cur_opcode in
    21) echo "   $cur_opcode: OP_Data$cur_opcode (= decimal $cur_opcode_dec)"
        op_data_show
        S20_OP_DATA33
        ;;
    41) echo "   $cur_opcode: OP_Data$cur_opcode (= decimal $cur_opcode_dec)"
        op_data_show
        S14_OP_DATA65
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 14 (OP_DATA65)         ###
#####################################
S14_OP_DATA65() {
  get_next_opcode
  case $cur_opcode in
    21) echo "   $cur_opcode: OP_Data$cur_opcode (= decimal $cur_opcode_dec)"
        op_data_show
        S15_OP_DATA33 
        ;;
    41) echo "   $cur_opcode: OP_Data$cur_opcode (= decimal $cur_opcode_dec)"
        op_data_show
        S17_OP_DATA65
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 15 (OP_DATA33)         ###
#####################################
S15_OP_DATA33() {
  get_next_opcode
  case $cur_opcode in
    52) echo "   $cur_opcode: OP_Data$cur_opcode (= decimal $cur_opcode_dec)"
        op_data_show
        S22_PRE_MULTISIG
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 17 (OP_DATA65)         ###
#####################################
S17_OP_DATA65() {
  get_next_opcode
  case $cur_opcode in
    41) echo "   $cur_opcode: OP_Data$cur_opcode (= decimal $cur_opcode_dec)"
        op_data_show
        S18_OP_DATA65
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 18 (OP_DATA65)         ###
#####################################
S18_OP_DATA65() {
  get_next_opcode
  case $cur_opcode in
    53) echo "   $cur_opcode: OP_Data$cur_opcode (= decimal $cur_opcode_dec)"
        S22_PRE_MULTISIG
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 20 (OP_DATA33)         ###
#####################################
S20_OP_DATA33() {
  get_next_opcode
  case $cur_opcode in
    21) echo "   $cur_opcode: OP_Data$cur_opcode (= decimal $cur_opcode_dec)"
        op_data_show
        S21_OP_DATA33
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 21 (OP_DATA33)         ###
#####################################
S21_OP_DATA33() {
  get_next_opcode
  case $cur_opcode in
    52) echo "   $cur_opcode: OP_Data$cur_opcode (= decimal $cur_opcode_dec)"
        op_data_show
        S22_PRE_MULTISIG
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 22 (PRE_MULTISIG)     ###
#####################################
S22_PRE_MULTISIG() {
  get_next_opcode
  case $cur_opcode in
    AE) echo "   $cur_opcode: OP_CHECKMULTISIG"
        S23_MULTISIG
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 23 (MULTISIG)          ###
#####################################
S23_MULTISIG() {
    echo "* This is a Multisig script"
}
#####################################
### STATUS 24 (OP_RETURN)         ###
#####################################
S24_OP_RETURN() {
  get_next_opcode
  case $cur_opcode in
    28) echo "   $cur_opcode: OP_Data$cur_opcode (= decimal $cur_opcode_dec)"
        offset=$(( offset + 1 ))
        op_data_show
        S25_NULL_DATA
        ;;
    *)  echo "   $cur_opcode: unknown opcode "
        ;;
  esac
}
#####################################
### STATUS 25 (NULL_DATA)         ###
#####################################
S25_NULL_DATA() {
    echo "* This is a NULLDATA script"
}
#####################################
### STATUS 26 (UNKNOWN)           ###
#####################################
S26_UNKNOWN() {
    echo "* This is an UNKNOWN script"
}

##########################################################################
### AND HERE WE GO ...                                                 ###
##########################################################################

typeset -i cur_opcode_dec
offset=0

opcode_array=$( echo $param | sed 's/[[:xdigit:]]\{2\}/ &/g' )
opcode_array_elements=$( echo ${#opcode_array} / 3 | bc )
# echo "opcode_array_elements=$opcode_array_elements, array=$opcode_array"

shell_string=$( echo $SHELL | cut -d / -f 3 )
if [ "$shell_string" == "bash" ] ; then
  i=0
  j=1
  declare -a opcode_ar
  while [ $i -lt $opcode_array_elements ]
   do
    # echo ${opcode_array:$j:2}
    opcode_ar[$i]=${opcode_array:$j:2}
    # echo "opcode_ar[$j]=$opcode_ar[$j]"
    i=$(( i + 1 ))
    j=$(( j + 3 ))
  done
elif [ "$shell_string" == "ksh" ] ; then
  set -A opcode_ar $opcode_array
fi

#####################################
### STATUS 0  INIT                ###
#####################################
  while [ $offset -lt $opcode_array_elements ]  
   do

    get_next_opcode

    case $cur_opcode in
      76) echo "   $cur_opcode: OP_DUP"
	  S1_OP_DUP
          ;;
      A9) echo "   $cur_opcode: OP_HASH160"
	  S8_OP_HASH160
          ;;
      51) echo "   $cur_opcode: OP_1, OP_TRUE"
	  S13_OP_1
          ;;
      6A) echo "   $cur_opcode: OP_RETURN"
	  S24_OP_RETURN
          ;;
      20) echo "   $cur_opcode: OP_Data32"
	  S26_UNKNOWN
          ;;
      21) echo "   $cur_opcode: OP_Data33"
          op_data_show
	  S11_OP_DATA33
          ;;
      24) echo "   $cur_opcode: OP_DATA36"
	  S26_UNKNOWN
          ;;
      41) echo "   $cur_opcode: OP_DATA65"
          op_data_show
	  S6_OP_DATA65
          ;;
      *)
          ;;
    esac

    if [ $offset -gt 100 ] ; then
      echo "emergency exit, output scripts should not reach this size ..."
      exit 1
    fi
  done

echo "$ret_string"


